aws-rds-dbinstance/src/main/java/software/amazon/rds/dbinstance/util/ImmutabilityHelper.java

```diff
@@ -0,0 +1,88 @@
+
+    private static final String AURORA = "aurora";
+    private static final String AURORA_MYSQL = "aurora-mysql";
+
+    protected static final List<String> DEPRECATED_ORACLE_ENGINES = Arrays.asList(ORACLE_SE, ORACLE_SE1);
+
+    private ImmutabilityHelper() {
+    }
+
+    static boolean isUpgradeToOracleSE2(final ResourceModel previous, final ResourceModel desired) {
+        return previous.getEngine() != null &&
+                DEPRECATED_ORACLE_ENGINES.contains(previous.getEngine().toLowerCase()) &&
+                ORACLE_SE2.equalsIgnoreCase(desired.getEngine());
+    }
+
+    static boolean isUpgradeToAuroraMySQL(final ResourceModel previous, final ResourceModel desired) {
+        return previous.getEngine() != null &&
+                desired.getEngine() != null &&
+                previous.getEngine().equals(AURORA) &&
+                desired.getEngine().equals(AURORA_MYSQL);
+    }
+
+    static boolean isAZMutable(final ResourceModel previous, final ResourceModel desired) {
+        return Objects.equal(desired.getAvailabilityZone(), previous.getAvailabilityZone()) ||
+                desired.getAvailabilityZone() == null &&
+                        Boolean.TRUE.equals(desired.getMultiAZ());
+    }
+
+    static boolean isPerformanceInsightsMutable(final ResourceModel previous, final ResourceModel desired) {
+        return previous.getEnablePerformanceInsights() == null ||
+                !previous.getEnablePerformanceInsights() ||
+                !desired.getEnablePerformanceInsights() ||
+                Objects.equal(previous.getPerformanceInsightsKMSKeyId(), desired.getPerformanceInsightsKMSKeyId());
+    }
+
+    static boolean isResourceModelMutable(final ResourceModel previous, final ResourceModel desired) {
+        return (new EqualsBuilder())
+                .append(desired.getDBClusterIdentifier(), previous.getDBClusterIdentifier())
+                .append(desired.getDBName(), previous.getDBName())
+                .append(desired.getMasterUsername(), previous.getMasterUsername())
+                .append(desired.getPort(), previous.getPort())
+                .append(desired.getDBSnapshotIdentifier(), previous.getDBSnapshotIdentifier())
+                .append(desired.getDBSubnetGroupName(), previous.getDBSubnetGroupName())
+                .append(desired.getSourceDBInstanceIdentifier(), previous.getSourceDBInstanceIdentifier())
+                .append(desired.getDBInstanceIdentifier(), previous.getDBInstanceIdentifier())
+                .append(desired.getPubliclyAccessible(), previous.getPubliclyAccessible())
+                .append(desired.getCharacterSetName(), previous.getCharacterSetName())
+                .append(desired.getStorageEncrypted(), previous.getStorageEncrypted())
+                .append(desired.getKmsKeyId(), previous.getKmsKeyId())
+                .append(desired.getTimezone(), previous.getTimezone())
+                .append(desired.getSourceRegion(), previous.getSourceRegion())
```

Comment:
Some of these attributes is available in ModifyDBInstance request i.e:DBSubnetGroupName, Port...etc . Why CFN can't change them by update handler and need replacement?

ref: https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_ModifyDBInstance.html

Reply:
Everything that's a part of modifyDBInstance request should remain there, in this case AWS RDS API will handle the modification in the best possible way. Hence, I'll be moving all modify-known attributes out of this check. Thank you for this finding!